### DISCUSSION TITLE
File Manager

### DISCUSSION THREAD

**Discussion by andrewg78** [BIPES/BIPES#85] [Q&A] [answered]
1. Does FileManager work on RaspberryPi (anyLinux)?
2. I can see only two files under "Templates' whereas there are several files in BIPES/ui/pylibs . Is this by design?
3. Upload does not push any files to rpi. Perhaps some folder permission. But I have no idea here it suppose to send the file.
4. Saving changes to any of these two python files from the Templates section does not work either.
My configuration:
 BIPES is served through the NGNIX on Raspberry Pi Zero W (headless mode). 
 I use websocket then to establish connection from my laptop
![filemanager](https://user-images.githubusercontent.com/2042344/147393181-fd9732b4-b901-425b-ba43-9280d3bcdde4.png)

**Comment by gastmaier** [comment, marked as answer]
Hi!
1 - The file manager does not work with raspberry pi (any linux), only with a device running MicroPython.
2 - pylibs includes MicroPython libraries that are implemented in the "Blocks" tab, for example, under Blocks > Displays > LED Matrix there is the "Install tm1640 library" button which installs the necessary library to use the TM1650 LED matrix.
3 - Uploading won't work with raspberry pi (any linux) due to (1).
4 - The Templates inside the Files tab is a forgotten section from early on versions of BIPES and don't do anything other than serving static files.

BIPES as is is not well compatible with linux, even though there were efforts to bridging the Python REPL to be accessed by BIPES, it is not very well tested and doesn't seem to be working with new versions of Python...

I personally suggest, when running a full pledge os like linux in the embedded device, use ssh, scp and even [syncthing](https://syncthing.net/) to share code to the device, since it has way more capability than a constrained device running MicroPython.

**Reply by gastmaier** [reply]
Just to wrap up,
if I were to implement full Linux support, I would do based on web-SSH to have full control over the embedded device and its files, then request run the python code on demand, instead of trying to expose just the Python REPL to BIPES.

**Comment by rafaelaroca** [comment]
Hello @andrewg78 and @JorgeGMarques 

The idea is to have the File Manager working with any Linux as soon as possible. Some features already work. Considering that you are connected to a Linux box using SocketServerWebREPL (https://github.com/BIPES/SocketServerWebREPL), as we have discussed on other forum thread and issues.

Well, after the WebREPL connection is active, you can:

List files on the remote Raspberry Pi / any Linux device, as seen here:
![image](https://user-images.githubusercontent.com/31751413/147395504-80f1aeb8-ff5d-4ce0-a60d-22b8468a84a3.png)

Run a remote Python file on the remote device:
![image](https://user-images.githubusercontent.com/31751413/147395509-ed4a5914-19b3-4833-8f29-8311cfb1b8ce.png)

Viewing, editing, and saving remote files should work, but there is some bug that we still did not figure out that is preventing this figure from working. Hope to get this working soon.

**Comment by andrewg78** [comment]
When I use refresh device file list button, I see non repetitive behavior under LInux console.
![refres-list-dir png](https://user-images.githubusercontent.com/2042344/147472850-23d62689-26fb-4e9b-9a15-51b833175198.png)

**Reply by rafaelaroca** [reply]
Hi, is socketserverWebREPL.py on the last version? Could you please try to run it with python2?

python2 socketserverWebREPL.py

Just tested it here and works fine. Trying to figure out your scenario...

regards.

**Comment by andrewg78** [comment]
Yes, it's python 2.7.16 and the newest socketserverWebREPL.py
In the webREPL folder I see a tmpCode.py file with the blocks converted to the code, so there is a proper connection.
The file was created by the root user.
My scenario it's a BIPES served through the NGINX on the RaspberryPi

**Comment by rafaelaroca** [comment]
It is strange because the output you showed is different than the output I have with the latest version. On my system, I see a lot of debug messages:

![image](https://user-images.githubusercontent.com/31751413/147476354-bfea8517-133f-4d13-899a-5ed85f593b12.png)

![image](https://user-images.githubusercontent.com/31751413/147476368-5cc7a64a-563d-458e-acf7-b89b24097482.png)

![image](https://user-images.githubusercontent.com/31751413/147476387-dba23d79-2403-47a9-acdd-dd6db9352b1d.png)

Serving BIPES using NGINX is perfect. No problem with that, as all BIPES files are static. All the REPL communication is using websockets.

About  tmpCode.py, it is okay. socketserverWebREPL.py creates this file with the contents of the file asked to be executed on the host device.

**Comment by andrewg78** [comment]
After the python shell is started (s key) I see debug also.
Could this be a timing issue? Some chars are not received
![debug](https://user-images.githubusercontent.com/2042344/147478098-19a39c62-2271-428e-9082-f27709ae484a.png)
2.
When I try to send a file I get an exception
![send-file-exception](https://user-images.githubusercontent.com/2042344/147478177-36e798c6-335f-4dee-9a7a-649b1c071479.png)

**Reply by rafaelaroca** [reply]
Hi! These messages are really for low-level debugging. What you see is correct as the network packets are received in "chunks". Note that the system receives "os.listdir" and after it receives "r()". Next, socketServerWebREPL assembles all received packets and we get: "os.listdir()" which is correctly executed. So, all chars are received, but on different packets.

So, point 1 is ok :-)

About point 2, sending files is still bugged. Bear with us while we try to solve this problem.  Anyway, as a temporary workaround, you can already click on the "run" button, which will send the block-based program to the board and save it as tmpCode.py.

**Comment by andrewg78** [comment]
We have a nice progress :)
BIPES receives all files now!!! Seems like the fix to os.listdir method did the job ! Super
![get-files-ok](https://user-images.githubusercontent.com/2042344/147572860-c72644d3-25e2-4f6e-b38b-a8e0a9b4a81f.png)
There is however another issue.
When I click to load the file, I get disconnected and the websocker server prints to Linux terminal
stdout send: exception while writing to socket: 'NoneType' object has no attribute 'sendall'

**Reply by rafaelaroca** [reply]
Great. Loading files is a bug still not solved, sorry. But you can already run programs. See my messages on the "issues".

**Comment by andrewg78** [comment]
LICENSE is marked as a 'directory' which is not true, it's a file
-rw-r--r--  1 pi   pi   35149 Dec 24 14:47 LICENSE

**Reply by rafaelaroca** [reply]
This is another bug. We will check it...

**Comment by andrewg78** [comment]
I wonder if we should not have only the python files listed here ( + folders)......
